### PR TITLE
Cluster Updates for Cloud [API-311,API-225]

### PR DESCRIPTION
**Pull request by zpqrtbnk**
A set of updates to Cluster code, in preparation for Cloud support implementation.

### CODE REVIEW

File: src/Hazelcast.Net/Clustering/ClusterConnections.cs [hunk 1]
```diff
@@ -529,14 +546,16 @@ private async Task<Attempt<MemberConnection>> ConnectFirstAsync(NetworkAddress a
         /// </remarks>
         private async Task<Attempt<MemberConnection>> EnsureConnectionAsync(MemberInfo member, CancellationToken cancellationToken)
         {
+            HConsole.WriteLine(this, $"Ensure {_clusterState.ClientName} is connected to {member.Id.ToShortString()} at {member.Address}");
+
```
Comment (alexb5dh) [inline src/Hazelcast.Net/Clustering/ClusterConnections.cs:550]:
```suggestion

```

File: src/Hazelcast.Net/Core/ClientVersion.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                if (_clientVersion != null) return _clientVersion;
+
+                var type = typeof (ClientVersion);
+                var assembly = type.Assembly;
+                var attribute = assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>();
+                if (attribute != null)
+                {
+                    var version = attribute.InformationalVersion;
+#if NETSTANDARD2_1
+                    var pos = version.IndexOf('+', StringComparison.OrdinalIgnoreCase);
```
Comment (alexb5dh) [inline src/Hazelcast.Net/Core/ClientVersion.cs:46]: So we probably can use `StringNetStandardExtensions.IndexOf` for 2.0 here?

File: src/Hazelcast.Net/Networking/AddressProvider.cs [hunk 3]
```diff
@@ -73,7 +74,13 @@ public AddressProvider(NetworkingOptions networkingOptions, ILoggerFactory logge
         /// <returns>All addresses.</returns>
         public IEnumerable<NetworkAddress> GetAddresses()
         {
-            return (_privateToPublic ??= _createMap()).Keys;
+            if (_privateToPublic == null)
+            {
+                _privateToPublic = _createMap();
+                if (_privateToPublic == null)
+                    throw new HazelcastException("Failed to obtain an address map.");
```
Comment (alexb5dh) [inline src/Hazelcast.Net/Networking/AddressProvider.cs:81]:
```suggestion
                _privateToPublic = _createMap() ?? throw new HazelcastException("Failed to obtain an address map.");
```

File: src/Hazelcast.Net/Networking/AddressProvider.cs [hunk 4]
```diff
@@ -90,6 +97,9 @@ public NetworkAddress Map(NetworkAddress address)
             if (_privateToPublic == null)
             {
                 _privateToPublic = _createMap();
+                if (_privateToPublic == null)
+                    throw new HazelcastException("Failed to obtain an address map.");
```
Comment (alexb5dh) [inline src/Hazelcast.Net/Networking/AddressProvider.cs:101]:
```suggestion
                _privateToPublic = _createMap() ?? throw new HazelcastException("Failed to obtain an address map.");
```

File: src/Hazelcast.Net/Clustering/ClusterMembers.cs [hunk 5]
```diff
@@ -253,6 +255,28 @@ public MembersUpdatedEventArgs SetMembers(int version, ICollection<MemberInfo> m
                     else diff[m] = 2;
             }
 
+            // log, if the members have changed (one of them at least is not 3=unchanged)
+            if (_logger.IsEnabled(LogLevel.Information) && diff.Any(d => d.Value != 3))
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net/Clustering/ClusterMembers.cs:259]: yes fixing

**Pull request merged**